Homework Task 5 (Романенко Денис)

NadirCianid opened the pull request.

NadirCianid committed:
Реализовал через две мапы.

NadirCianid committed:
Merge branch 'polis-vk:master' into homework_task_4

NadirCianid committed:
исправил проблемы

NadirCianid committed:
исправил замечания

Избавился от мап для сортированных ключей/значений, перейдя на стримы.

NadirCianid committed:
Реализовал WordFrequency

NadirCianid committed:
Реализовал MailService и прилагающиеся классы

Пример работы сервиса в классе MailDriver

11cfif commented on src/main/java/ru/mail/polis/homework/collections/mail/MailDriver.java:
```diff
@@ -0,0 +1,33 @@
+
+        MailService.process(mailService, mailList);
+
+        System.out.println(mailService.getMailBox());
```
ну вот получили вы почту свою и хотите извлечь содержимое. И только с помощью метода туСтринг?? туСтринг вообще обычно используется сугубо для отладочныъ целей (чтобы в логах было хорошо видно и так далее). Получается мне пришли письма, а что в них (контент) я не могу получить напрямую? надо делать какие то парсинги

11cfif commented on src/main/java/ru/mail/polis/homework/collections/mail/MailDriver.java:
```diff
@@ -0,0 +1,33 @@
+import java.util.List;
+
+/**
+ * Пример работы сервиса
```
для такого лучше тест делать ;-)

11cfif commented on src/main/java/ru/mail/polis/homework/collections/mail/MailService.java:
```diff
@@ -1,58 +1,74 @@
  */
-public class MailService implements Consumer {
+public class MailService<M extends Mail> implements Consumer<M> {
+    private final Map<String, List<M>> mailBox = new HashMap<>();
```
3 мапы! можно 2 -мя обойтись

11cfif commented on src/main/java/ru/mail/polis/homework/collections/mail/MailService.java:
```diff
@@ -1,58 +1,74 @@
      */
     public String getPopularSender() {
-        return null;
+        return sendersByPopularity.entrySet().stream()
```
условие задания прочитайте. Зачем что то реализовывать, что было уже реализовано? Да и еще с ошибками из прошлой дз

11cfif commented on src/main/java/ru/mail/polis/homework/collections/streams/WordFrequency.java:
```diff
@@ -21,7 +26,16 @@ public class WordFrequency {
-        return null;
+        return lines.
+                flatMap(s -> Arrays.stream(s.split("[ .,!:-?;]")))
+                .filter(s -> !s.equals(""))
```
блин, ну сколько можно ! нельзя так на пустоту проверять

NadirCianid committed:
Merge branch 'homework_task_4' into homework_task_5

NadirCianid committed:
Исправил замечания

Встроил работу с популярностью через PopularMap. Написал тест вместо класса MailDriver.

NadirCianid committed:
Добавил метод getContent в классы почты, чтобы можно было извлечь содержимое.

11cfif commented on build.gradle:
```diff
@@ -5,7 +5,7 @@ plugins {
 version '1.0-SNAPSHOT'
 
-sourceCompatibility = 1.8
+sourceCompatibility = 8
```
?

11cfif commented on src/main/java/ru/mail/polis/homework/collections/mail/Mail.java:
```diff
@@ -3,5 +3,28 @@
+        return recipient;
+    }
+    public String getContent() {
+        return "";
```
??

11cfif commented on src/main/java/ru/mail/polis/homework/collections/mail/Salary.java:
```diff
@@ -3,5 +3,21 @@
+
+    @Override
+    public String getContent() {
+        return String.valueOf(sum);
```
зарплату только в строке можно получить? ну как так то

11cfif replied:
юзайте дженерики

11cfif commented on src/main/java/ru/mail/polis/homework/collections/mail/MailService.java:
```diff
@@ -1,58 +1,68 @@
 
+        List<M> newMail = new ArrayList<>();
+        newMail.add(mail);
+        mailBox.merge(mail.getRecipient(), newMail, (currentList, additionalPart) -> Stream.concat(currentList.stream(), additionalPart.stream())
```
ох) изучайте дальше методы мапы. там есть спец методы для мап с коллекциями внутри

11cfif commented on src/main/java/ru/mail/polis/homework/collections/streams/WordFrequency.java:
```diff
@@ -21,7 +24,16 @@ public class WordFrequency {
     public static List<String> wordFrequency(Stream<String> lines) {
-        return null;
+        return lines.
+                flatMap(s -> Arrays.stream(s.split("[ .,!:-?;]")))
```
точку потеряли

NadirCianid committed:
Реализовал метод getContent через дженерик.

NadirCianid committed:
упростил добавление новой почты

11cfif commented on build.gradle:
```diff
@@ -17,3 +17,4 @@ dependencies {
     testCompile group: 'commons-io', name: 'commons-io', version: '2.6'
 
 }
+targetCompatibility = JavaVersion.VERSION_1_8
```
лишние файлы

11cfif commented on src/main/java/ru/mail/polis/homework/collections/mail/Mail.java:
```diff
@@ -3,5 +3,31 @@
+    }
+
+    public C getContent() {
+        return content;
```
переменная которая нигде не используется

11cfif commented on src/main/java/ru/mail/polis/homework/collections/mail/MailMessage.java:
```diff
@@ -3,6 +3,21 @@
  */
-public class MailMessage {
+public class MailMessage extends Mail<String> {
+    private final String message;
```
зачем то дублирует контент

11cfif commented on src/main/java/ru/mail/polis/homework/collections/mail/Salary.java:
```diff
@@ -3,5 +3,21 @@
  */
-public class Salary {
+public class Salary extends Mail<Integer> {
+    private final int sum;
```
аналогично

11cfif commented on src/main/java/ru/mail/polis/homework/collections/mail/MailService.java:
```diff
@@ -1,58 +1,67 @@
  * Всего 7 тугриков за пакет mail
  */
-public class MailService implements Consumer {
+public class MailService<M extends Mail> implements Consumer<M> {
```
Mail не дженерифицирован

11cfif commented on src/main/java/ru/mail/polis/homework/collections/mail/MailService.java:
```diff
@@ -1,58 +1,67 @@
 
+        List<M> newMail = new ArrayList<>();
+        newMail.add(mail);
+        mailBox.merge(mail.getRecipient(), newMail, (currentList, additionalPart) -> {
```
метод мапы есть для коллекций

11cfif commented on src/main/java/ru/mail/polis/homework/collections/mail/MailService.java:
```diff
@@ -1,58 +1,67 @@
      */
-    public static void process(MailService service, List mails) {
-
+    public static void process(MailService service, List<? extends Mail> mails) {
```
MailService не джеренефицирован

11cfif reviewed (commented):
на лекции разбирал домашку
